lab_3/main.py

```diff
@@ -17,38 +19,76 @@ def tokenize_by_sentence(text: str) -> tuple:
          (('_', 'h', 'e', '_'), ('_', 'i', 's', '_'), ('_', 'h', 'a', 'p', 'p', 'y', '_'))
          )
     """
-    pass
+    if not isinstance(text, str):
+        return ()
+
+    sentences = re.split(r'\W{1,3}\s(?=[A-Z])', text)
+    result = []
+
+    for sentence in sentences:
+        sent = []
+        for token in re.sub('[^a-z ]', '', sentence.lower()).split():
+            word = []
+            for index, letter in enumerate(token):
+                if index == 0:
+                    word += '_' + letter
+                elif index == len(token) - 1:
+                    word += letter + '_'
+                    sent.append(tuple(word))
+                else:
+                    word += letter
+        if sent:
+            result.append(tuple(sent))
+
+    return tuple(result)
 
 
 # 4
 class LetterStorage:
 
     def __init__(self):
-        pass
+        self.storage = {}
 
     def _put_letter(self, letter: str) -> int:
         """
         Puts a letter into storage, assigns a unique id
         :param letter: a letter
         :return: 0 if succeeds, 1 if not
         """
-        pass
+        if not isinstance(letter, str):
+            return 1
+
+        if letter not in self.storage:
+            letter_id = 0
+            self.storage[letter] = letter_id
+            letter_id += 1
```

Comment:
This `letter_id` is reset to `0` every time this method is called. Find a way to use `letter_id` on the class-level.

Reply:
got it!